### PR TITLE
Carbon Calculator Actions update - three issues

### PR DESCRIPTION
**Pull request by BradHN1**
####  Summary / Highlights
Recent Carbon Calculator updates introduced some database errors, unlinking Actions and Calculator Actions.

#### Details (Give details about what this PR accomplishes, include any screenshots etc)
This PR address three issues:

#808: Importing from CSV breaks Action/Calculator Action linkage: when the Carbon Calculator is updated or the input data changes, the code automatically imports those changes.  There was a bug which for existing Calculator Actions it was deleting the old one and creating a new one, instead of updating the old one.  That broke the linkage between actions and calculator actions, so carbon reduction values in dev and canary all got set to zero.

#809: fix initialization error - message about duplicate default values
Another bug fix: for a given variable, location and valid_date there should be only one CalcDefault value in the database, but there were duplicates, causing a message.  Added a utility routine when the Calculator is updated which removes those duplicate values. 

#810: add new actions and link all Actions to Calculator Actions
First I created a number of new actions in AirTable, exporting the expanded list to the file Actions.csv
I downloaded the list of all actions in all communities for prod, canary, dev and my local db, merged them, eliminated duplicates, and assigned a carbon calculator action to each one.  About 2000 lines, a lot of work.

A task 'update-actions-content' was added, which if feature flag update-actions-content is set goes thorugh the complete list, finds the action in the database, updates the calculator_action.
If the calculator_action is set to "DELETE", then the action is marked as deleted.

No database migrations are required

#### Testing Steps (Provide details on how your changes can be tested)
After deploying to dev/canary/prod or your local database, create an admin task, called update-actions-content, and a feature flag.  Run the task as a one-off and all the Actions will be fixed.

#### Requirements (place an `x` in each `[ ]`)
* [x] I've read and understood the [Contributing Guidelines](/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've tested my changes manually.
* [ ] I've added unit tests to my PR.

##### Transparency ([Project board](https://github.com/orgs/massenergize/projects/7/views/2))
* x] I've given my PR a meaningful title.
* [x] I linked this PR to the relevant issue.
* [ x] I moved the linked issue from the "Sprint backlog" to "In progress" when I started this.
* [x ] I moved the linked issue to "QA Verification" now that it is ready to merge.

### CODE REVIEW

File: src/carbon_calculator/CCDefaults.py [hunk 1]
```diff
@@ -25,56 +22,63 @@ def getLocality(inputs):
...
-                    DefaultsByLocality[c.locality][c.variable]["values"].append(c.value)
-
-
-    except Exception as e:
-        print(str(e))
-        print("CalcDefault initialization skipped")
 
-    def __init__(self):
-        print("CCD __init__ called")
+    def ready(self):
```
Comment (Opoku-Agyemang) [inline src/carbon_calculator/CCDefaults.py:37]: what does ready do?  Can we add some comments to explain what it does?

File: src/carbon_calculator/CCDefaults.py [hunk 2]
```diff
@@ -25,56 +22,63 @@ def getLocality(inputs):
...
+                            # insert value at this point
+                            f = True
+                            self.DefaultsByLocality[c.locality][c.variable]["valid_dates"].insert(i,date)
+                            self.DefaultsByLocality[c.locality][c.variable]["values"].insert(i,c.value)
+                            break
+                        elif date == valid_date:
+                            # multiple values with one date; clean this from database
+                            print("Duplicate encountered")
+                            f = True
+                            break
```
Comment (Opoku-Agyemang) [inline src/carbon_calculator/CCDefaults.py:66]: this function is long and cumbersome with no clear explantion of what is happening.  
It will be extremely difficult to test and quite tidious to follow and give a proper review.  

I would appreciate it if we split it into smaller helper functions with good names (and comments) that get called in this main one.

We will be able to test the different aspects of it better that way.

File: src/carbon_calculator/CCDefaults.py [hunk 3]
```diff
@@ -119,15 +125,19 @@ def exportDefaults(self,fileName):
...
         try:
             status = True
             with open(fileName, newline='') as csvfile:
                 inputlist = csv.reader(csvfile)
                 first = True
+                num = 0
                 for item in inputlist:
-                    if first:
+                    if first:   # header row
                         t = {}
```
Comment (Opoku-Agyemang) [inline src/carbon_calculator/CCDefaults.py:141]: it is probably a better idea to not name variables with one letter.   It makes it hard to follow or even guess what it contains.

Also first is also weird.  Would you go for something like `is_first_run` ?

File: src/carbon_calculator/carbonCalculator.py [hunk 4]
```diff
@@ -190,6 +194,10 @@ def __init__(self, reset=False) :
                 theInstance = theClass(name)
                 self.allActions[name] = theInstance
 
+            # uncomment to measure elapsed time
+            #end = time.time()
+            #print(end - start)
```
Comment (Opoku-Agyemang) [inline src/carbon_calculator/carbonCalculator.py:199]: remove all unnecessary commented out code

File: src/carbon_calculator/carbonCalculator.py [hunk 5]
```diff
@@ -292,39 +300,80 @@ def ImportQuestions(self, questionsFile):
...
+                        #    category=item[1],
+                        #    question_text=item[2],
+                        #    question_type=item[3],
+                        #    response_1=item[4], skip_1=skip[0],
+                        #    response_2=item[6], skip_2=skip[1],
+                        #    response_3=item[8], skip_3=skip[2],
+                        #    response_4=item[10], skip_4=skip[3],
+                        #    response_5=item[12], skip_5=skip[4],
+                        #    response_6=item[14], skip_6=skip[5])
+                        #print('Importing Question ',question.name,': ',question.question_text)
```
Comment (Opoku-Agyemang) [inline src/carbon_calculator/carbonCalculator.py:360]: remove

File: src/carbon_calculator/carbonCalculator.py [hunk 6]
```diff
@@ -292,39 +300,80 @@ def ImportQuestions(self, questionsFile):
...
+                        #    if len(item[16]):
+                        #        question.minimum_value = eval(item[16])
+                        #    if len(item[17])>0:
+                        #        question.maximum_value = eval(item[17])
+                        #    if len(item[18])>0:
+                        #        question.typical_value = eval(item[18])
+
+                        #question.save()
+                        #num+=1
+
```
Comment (Opoku-Agyemang) [inline src/carbon_calculator/carbonCalculator.py:372]: remove

File: src/carbon_calculator/carbonCalculator.py [hunk 7]
```diff
@@ -343,31 +392,51 @@ def ImportActions(self, actionsFile):
...
+                        #        picture = SavePic2Media(item[t["Picture"]])
+                        #        action = Action(name=item[0],
+                        #            title = item[t["Title"]],
+                        #            description=item[t["Description"]],
+                        #            helptext=item[t["Helptext"]],
+                        #            category=item[t["Category"]],
+                        #            average_points=int(eval(item[t["Avg points"]])),
+                        #            questions=item[t["Questions"]].split(","),
+                        #            picture = picture)
+                        #        action.save()
```
Comment (Opoku-Agyemang) [inline src/carbon_calculator/carbonCalculator.py:434]: remove

**Review by Opoku-Agyemang — CHANGES_REQUESTED** [review]

File: src/carbon_calculator/CCDefaults.py (hunk 1 above)
Comment (BradHN1) [reply src/carbon_calculator/CCDefaults.py:37]: done

File: src/carbon_calculator/CCDefaults.py (hunk 2 above)
Comment (BradHN1) [reply src/carbon_calculator/CCDefaults.py:66]: cleaned.  comments attempt to explain it.
the routine has been exhaustively tested and used for years, don't have time to rewrite it now.

File: src/carbon_calculator/CCDefaults.py (hunk 3 above)
Comment (BradHN1) [reply src/carbon_calculator/CCDefaults.py:141]: done

File: src/carbon_calculator/carbonCalculator.py (hunk 5 above)
Comment (BradHN1) [reply src/carbon_calculator/carbonCalculator.py:360]: done

File: src/carbon_calculator/carbonCalculator.py (hunk 7 above)
Comment (BradHN1) [reply src/carbon_calculator/carbonCalculator.py:434]: done

File: src/carbon_calculator/carbonCalculator.py (hunk 6 above)
Comment (BradHN1) [reply src/carbon_calculator/carbonCalculator.py:372]: done

File: src/carbon_calculator/carbonCalculator.py (hunk 4 above)
Comment (BradHN1) [reply src/carbon_calculator/carbonCalculator.py]: done
